File: c++/sumOfLeftLeaves.cpp
//
//  sumOfLeftLeaves.cpp
//  codefights
//
//  Created by YANGXIAOJUN on 2017/1/7.
//  Copyright © 2017年 YANGXIAOJUN. All rights reserved.
//  Leetcode: 404. Sum of Left Leaves

#include  <cstdio>

struct TreeNode {
    int val;
    TreeNode *left;
    TreeNode *right;
    TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 } ;

class Solution {
public:
    int sumOfLeftLeaves(TreeNode* root) {
        int ans = 0 ;
        
        if(root)
        {
            TreeNode* left = root->left ;
            TreeNode* right = root->right ;
            
            if(left && (left->left == nullptr && left->right == nullptr))
            {
                ans += left->val ;
            }
            
            ans += sumOfLeftLeaves(left) + sumOfLeftLeaves(right) ;
         }
        
        return 0 ;
    }
};

File: lintcode/src/main/java/com/vkeonline/lintcode/SumOfLeftLeaves.java
package com.vkeonline.lintcode;

import com.vkeonline.lintcode.common.TreeNode;

public class SumOfLeftLeaves {

    public int sumOfLeftLeaves(TreeNode root) {
        if(root == null) {
            return 0 ;
        }

        int sum = 0 ;

        if(root.left != null) {
            TreeNode left = root.left ;
            if(left.left == null && left.right == null)
                sum += left.val ;
            else {
                sum += sumOfLeftLeaves(left) ;
            }
        }

        if(root.right != null) {
            sum += sumOfLeftLeaves(root.right) ;
        }

        return sum ;

    }
}
